service/history/api/updateactivityoptions/api.go

```diff
@@ -0,0 +1,266 @@
+			mutableState := workflowLease.GetMutableState()
+			return updateActivityOptions(shardContext, validator, mutableState, request, response)
+		},
+		nil,
+		shardContext,
+		workflowConsistencyChecker,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return response, err
+}
+
+func updateActivityOptions(
+	shardContext shard.Context,
+	validator *api.CommandAttrValidator,
+	mutableState workflow.MutableState,
+	request *historyservice.UpdateActivityOptionsRequest,
+	response *historyservice.UpdateActivityOptionsResponse,
+) (*api.UpdateWorkflowAction, error) {
+	if !mutableState.IsWorkflowExecutionRunning() {
+		return nil, consts.ErrWorkflowCompleted
+	}
+	updateRequest := request.GetUpdateRequest()
+	mergeFrom := updateRequest.GetActivityOptions()
+	activityId := updateRequest.GetActivityId()
+
+	ai, activityFound := mutableState.GetActivityByActivityID(activityId)
+
+	if !activityFound {
+		// Looks like ActivityTask already completed as a result of another call.
+		// It is OK to drop the task at this point.
+		return nil, consts.ErrActivityTaskNotFound
+	}
+	mask := updateRequest.GetUpdateMask()
+	if mask == nil {
+		return nil, serviceerror.NewInvalidArgument("UpdateMask is nil")
+	}
+
+	updateFields := util.ParseFieldMask(mask)
+	mergeInto := &activitypb.ActivityOptions{
+		TaskQueue: &taskqueuepb.TaskQueue{
+			Name: ai.TaskQueue,
+		},
+		ScheduleToCloseTimeout: ai.ScheduleToCloseTimeout,
+		ScheduleToStartTimeout: ai.ScheduleToStartTimeout,
+		StartToCloseTimeout:    ai.StartToCloseTimeout,
+		HeartbeatTimeout:       ai.HeartbeatTimeout,
+		RetryPolicy: &commonpb.RetryPolicy{
+			BackoffCoefficient: ai.RetryBackoffCoefficient,
+			InitialInterval:    ai.RetryInitialInterval,
+			MaximumInterval:    ai.RetryMaximumInterval,
+			MaximumAttempts:    ai.RetryMaximumAttempts,
+		},
+	}
+
+	// update activity options
+	err := applyActivityOptions(mergeInto, mergeFrom, updateFields)
+	if err != nil {
+		return nil, err
+	}
+
+	// validate the updated options
+	err = validateActivityOptions(validator, request.NamespaceId, ai, mergeInto)
+	if err != nil {
+		return nil, err
+	}
+
+	// move forward activity version
+	ai.Stamp++
+
+	// invalidate timers
+	ai.TimerTaskStatus = workflow.TimerTaskStatusNone
+
+	// regenerate retry tasks
+	if workflow.GetActivityState(ai) == enumspb.PENDING_ACTIVITY_STATE_SCHEDULED {
+		// two options - it can be in backoff, or waiting to be started
+		now := shardContext.GetTimeSource().Now()
+		if now.After(ai.ScheduledTime.AsTime()) {
+			// activity is past its scheduled time and ready to be started
+			// we don't really need to do generate timer tasks, it should be done in closeTransaction
+		} else {
+			// activity is in backoff
+			_, err = mutableState.RetryActivity(ai, nil)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// fill the response
+	response.ActivityOptions = mergeInto
+
+	return &api.UpdateWorkflowAction{
+		Noop:               false,
+		CreateWorkflowTask: false,
```

Comment:
nit: these are defaults, no need to specify them explicitly.

Reply:
This is how it is in the rest of the code, which I think is for readability. It is better to be explicit with what function is returning. Compiler will take care of it.